### PR TITLE
Allow passing in of monitor resources

### LINKED ISSUE #339: New-CmAzCoreMonitor - Allow importing of Storage, Log Analytics and AppInsights.
There will be scenarios where Cloudmarque will be required to integrate with existing solutions that contain App Insights, Log Analytics Workspaces and/or Storage Account. We want to be able to specify if we want to create new or link to existing.

The changes required are: 

* Amend Schema to allow optional publish service values of the above three resources in settings.
* Add the ability to pass in the dependency "cm-service" value of the above three resources.
* If both of the above values exist in settings, default to looking for existing resources through the dependency service value.
* Add validation for when the dependency value is populated but nothing is found, even if the publish service value is populated in settings.
* Validation that either or of these service values need to be populated.

This task has a dependency for #335 to be completed first.

### PR DESCRIPTION
**Pull request by tm-luke**
Closes #339

### CODE REVIEW

**Review by deepend-dev — DISMISSED** [review]

File: Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.json [hunk 1]
```diff
@@ -46,15 +49,22 @@
...
+        },
         "WorkspaceName": {
             "type": "String"
         },
         "Workbook": {
             "type": "object"
         }
     },
+    "variables": {
+        "workspaceResourceId": "[resourceid('Microsoft.OperationalInsights/workspaces', parameters('workspaceName'))]"
```
Comment (deepend-dev) [inline Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.json:63]: I think you should also use resource group in which workspace is deployed

File: Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1 [hunk 2]
```diff
@@ -181,46 +180,81 @@ function New-CmAzCoreMonitor {
...
-						@{ name = "insights-logs-networksecuritygroupflowevent"}
-					)
-				})
+	
+				New-CmAzIaasStorage -SettingsObject $storageObject -OmitTags
+
+				$storageService = $SettingsObject.service.publish.storage
 			}
+			else {
+				$storageService = $SettingsObject.service.dependencies.storage
```
Comment (deepend-dev) [inline Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1:222]: verbose for existing storage?

File: Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1 [hunk 3]
```diff
@@ -181,46 +180,81 @@ function New-CmAzCoreMonitor {
...
+						dependencies = @{
+							resourceGroup = $SettingsObject.service.publish.resourceGroup
+						};
+						publish = @{
+							storage = $SettingsObject.service.publish.storage
+						}
 					}
+					storageAccounts = @(@{
+						storageAccountName = $SettingsObject.name;
+						accountType = "Standard";
```
Comment (deepend-dev) [inline Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1:207]: I think use of semicolon can be made consistent

File: Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1 [hunk 4]
```diff
@@ -181,46 +180,81 @@ function New-CmAzCoreMonitor {
...
+						blobContainer = @(
+							@{ name = "insights-logs-addonazurebackuppolicy"},
```
Comment (deepend-dev) [inline Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1:209]: space between "}

File: Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1 [hunk 5]
```diff
@@ -181,46 +180,81 @@ function New-CmAzCoreMonitor {
...
+	
+				New-CmAzIaasStorage -SettingsObject $storageObject -OmitTags
+
+				$storageService = $SettingsObject.service.publish.storage
 			}
+			else {
+				$storageService = $SettingsObject.service.dependencies.storage
+			}
+
+			$storageAccountId = (Get-CmAzService -Service $storageService -ThrowIfUnavailable -ThrowIfMultiple).Id
```
Comment (deepend-dev) [inline Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1:225]: should there be location option to search storage? considering there can be core storage in multiple regions

File: Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1 [hunk 6]
```diff
@@ -181,46 +180,81 @@ function New-CmAzCoreMonitor {
...
 
-			New-CmAzIaasStorage -SettingsObject $storageObject -OmitTags
+			if ($SettingsObject.service.publish.workspace) {
+				$workspaceName = Get-CmAzResourceName -Resource "LogAnalyticsWorkspace" -Architecture "Core" -Location $SettingsObject.location -Name $SettingsObject.name
+			}
+			else {
+				Write-Verbose "Service locating existing log analytics workspace instance..."
+				$workspaceName = (Get-CmAzService -Service $SettingsObject.service.dependencies.workspace -ThrowIfUnavailable -ThrowIfMultiple).resourceName
```
Comment (deepend-dev) [inline Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1:232]: location option?

File: Cloudmarque.Azure/Resources/Project/Integration/monitor.yml [hunk 7]
```diff
@@ -4,21 +4,24 @@ name: coretestmonitor042021
 location: UK South
 advisorLowCPUThresholdPercentage: 5									                  # Can be set to 0, 5, 10, 15 or 20
 workspaceDataRetentionInDays: 30                                      # Optional
-
+storageDataRetentionInDays: 125                                       # Defaults to indefinite data retention when nothing or so is provided.
 service:
```
Comment (deepend-dev) [inline Cloudmarque.Azure/Resources/Project/Integration/monitor.yml:8]: leaving a line will look good here

File: Cloudmarque.Azure/Resources/Project/Integration/monitor.yml [hunk 8]
```diff
@@ -4,21 +4,24 @@ name: coretestmonitor042021
...
     solution: Core-Test-Sol
-    storage: Core-Test-St
-    workspace: Core-Test-Ws
     workbook: Core-Test-wbk
+    appInsights: Core-Test-Ai                                         # Can use publish or dependency variants of these values.
+    storage: Core-Test-St                                             # Can use publish or dependency variants of these values.
+    workspace: Core-Test-Ws                                           # Can use publish or dependency variants of these values.
   dependencies:
     keyvault: Core-Test-Kv
+    appInsights: Core-Test-Ai                                         # Can use publish or dependency variants of these values.
```
Comment (deepend-dev) [inline Cloudmarque.Azure/Resources/Project/Integration/monitor.yml:22]: would be worth mentioning a note which one takes priority

File: Cloudmarque.Azure/Resources/Schema/New-CmAzCoreMonitor.Schema.json [hunk 9]
```diff
@@ -46,6 +46,15 @@
             "maximum": 90,
             "default": 90
         },
+        "storageDataRetentionInDays": {
+            "$id": "#/workspaceDataRetentionInDays",
```
Comment (deepend-dev) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzCoreMonitor.Schema.json:50]: storageDataRetentionInDays

File: Cloudmarque.Azure/Resources/Schema/New-CmAzCoreMonitor.Schema.json [hunk 10]
```diff
@@ -46,6 +46,15 @@
...
+            "type": ["integer", "null"],
+            "title": "WorkspaceDataRetentionInDays",
```
Comment (deepend-dev) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzCoreMonitor.Schema.json:52]: storageDataRetentionInDays

**Review by deepend-dev — CHANGES_REQUESTED** [review]

File: Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1 (hunk 6 above)
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1:232]: Bear in mind, we are trying to link to an existing environment with minimal setup, so id argue that the service tag is enough to identify the resource we want in this case.

File: Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1 (hunk 5 above)
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.ps1:225]: Bear in mind, we are trying to link to an existing environment with minimal setup, so id argue that the service tag is enough to identify the resource we want in this case.

File: Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.json (hunk 1 above)
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/Core/monitoring/New-CmAzCoreMonitor.json:63]: See below

**Review by deepend-dev — APPROVED** [review]

**Pull request merged**